VC Issuer Guide for tbDEX (#1111)
TBD54566975/developer.tbd.website: site/docs/tbdex/issuer/quickstart.mdx

Reviewer: can you rename this file to `onboarding.mdx`?

# Onboarding as a tbDEX Issuer

Issuers are a critical component of the tbDEX ecosystem that ensure trust between Wallets and PFIs by independently verifying claims about the party to whom they are issuing the credential; a typical example of a claim in tbDEX is verifying that a Wallet owner is not on a sanctions list. Although issuance is not a concept that is exclusive to tbDEX - it can be applied across [Web5](/docs/) in general -, as a tbDEX issuer you'll apply the same concepts to your app. The goal of this guide is to serve as a map to existing documentation we've provided for developers using our Web5 SDK, as well as provide you with information specific to your use case as a tbDEX developer. In this guide, we'll cover a basic implementation of an express server using the `@web5/credentials` class to issue your own VC.

> ## Background

Author: I believe your other comments overlapped with this so we should be good to resolve this one

> ## Background

Reviewer: looking at the deploy preview, this section doesn't add any additional value and can be removed completely

https://deploy-preview-1111--tbd-website-developer.netlify.app/docs/tbdex/issuer/quickstart#background

Additionally, developing an Issuer app will require use of our `@web5/credentials` package via `npm`, which we'll cover in more specifics later.

## Overview

A credential issuer app runs as an API service that exposes endpoints where third parties can request credentials. You can see a full end-to-end implementation of a credential issuer's workflow [in our documentation here](/docs/web5/build/verifiable-credentials/fan-club-vc). For the sake of this section, we'll go over what you as a tbDEX issuer will need to set up your app:

- Create a web server using a technology like ** [express](https://expressjs.com/) ** that will serve as an API endpoint.
- Design the credential you'd like to offer.
- Implement your issuance path, where you'll accept a request for issuance, perform checks required to verify the requester's information, and mint the credential.
- Return the credential to the requester.

## Environment Setup

You'll perform the following commands to install the packages required for your Issuer:

```bash
npm i @web5/credentials
npm i express
```

Within your files, you'll use these imports:

> ```typescript

Author: Updated

import { VerifiableCredential } from "@web5/credentials";
import { api } from './api.js'
```

Additionally, you'll want to ensure that you create a Decentralized Identifier (DID), which will serve as your Issuer's identity, following the [instructions outlined here](/docs/web5/build/decentralized-identifiers/how-to-create-did#create-a-did-manually). You can find a sample code snippet below:

> ```typescript

Author: Removed

import { DidKeyMethod } from '@web5/dids';

const issuerDid = await DidKeyMethod.create();
```

> Note that the method above does require installing the `@web5/dids` package.

Author: Removed

> With all of this, you can start up your basic express API service in a `main.ts` file that looks as follows:

Reviewer: couple of unanswered questions here

> With all of this, you can start up your basic express API service in a `main.ts` file that looks as follows:

Author: Updated all to main.js (switched of ts to js as well)

```typescript
import { api } from './api.js'

const server = api.listen(config.port, () => {
  log.info(`Server listening on port ${config.port}`)
})

api.get('/credential', async (req, res) => {
    // Implement the /credential API route
 });
```

With this basic structure in place, you'll be ready to now design your credential and implement your issuance path. If you'd like to view a full, runnable project that implements this, see our [VC Issuer GitHub Example](https://github.com/TBD54566975/workshop-tbdex-abc-vc-issuer/tree/main).

## Design Your Credential

With your app set up to accept incoming routes, you can then design a model file to represent your VC. As a tbDEX-focused Issuer a typical example for issuance would be that you are issuing a credential that certifies that a person isn't on a sanctions list. You could model that class as follows:

```typescript
import { VerifiableCredential, PresentationExchange } from "@web5/credentials";

class SanctionsCredential {
  constructor(listsChecked) {
    // indicates the sanctions lists checked against
    this.listsChecked = listsChecked; 
  }
}
```

> In our example above, we can offer a class property that stores all the sanctions lists checked against to issue the credential. You have wide flexibility with what you can store in your credential class, which means that you're free to include any public attributes that may be relevant to the credential you are verifying.

Reviewer: ```suggestion
In the example above, you provide a class property that stores all the sanctions lists checked against to issue the credential. You have wide flexibility with what you can store in the credential model class, which means that you're free to include any public attributes that may be relevant to the credential you are issuing.
```

> Your credential model is ** not ** the place to store private information about the user you are issuing the credential to, such as banking information or other PII. 

Reviewer: do you mean the model class or the credential itself?

Author: Updated with clarifying text

Reviewer: ok this is not true. in fact, we're specifically gonna need KYC credentials that will store private PII info. I recommend just removing this sentence altogether.

Author: Going to remove - just so I can better understand, how is it that credentials can store PII without exposing that? If the credential is being passed around don't we want to remove PII and just let the VC stand for the claim?

## Implement Issuance

With your web server structure in place and credential designed, you're now ready to build out your issuance codepath. 

> ** BLOCKED ON THIS PART TO SHOW HOW TO ACTUALLY TAKE IN INFO TO BE VERIFIED **

Reviewer: I think there's some misunderstanding here. I've removed the word "verify" a couple of times throughout the guide. A verifier has a very distinct meaning in the context of VCs, and is a different entity than the issuer. A verifier is someone you give an issued credential to.

The issuer is the one making claims. They wouldn't necessarily take any info and verify it. They'd do their own checks (check a real Sanctions list, or check an internal system to make sure the person is an employee, or actually earned a degree from their university). The issuer is the one qualified to make to the claim

Reviewer: Now what you might want to highlight is the act of a applying for or requesting a credential from an issuer

Author: Thanks for clarifying - I updated the language to be more clear about the difference between verify and check, and this should help devs understand it.

Once you've performed any verification tasks within your own business logic, you can mint the VC:

```typescript
api.get('/credential', async (req, res) => {
    try {
        // Intake user data or send back any responses/redirects requesting more info
        const authHeader = req.headers['authorization'];

        if (!authHeader) {
            return res.status(401).json({ errors: ['Authorization header required'] });
        }

        const [_, compactJwt] = authHeader.split('Bearer ');
        const { signer, payload } = await JwtVerifier.verify(compactJwt);

        // Process & verify user data
        const listsChecked = verifyUserData(payload);

        if (listsChecked == null) {
            res.status(401).json({
                errors: "User failed sanctions Check."
            })
        }

>         // Mint the credential

Reviewer: ```suggestion
        // Create the credential
```

        const vc = await VerifiableCredential.create({
            type: "SanctionsCredential",
            issuer: issuerDid.did,
            subject: signer,
            data: new SanctionsCredential(listsChecked)
        });
    }
 });
```

> Now the `vc` object you have at the end of this snippet will be a raw VC that looks like this:

Reviewer: ```suggestion
The `vc` object that's returned will be a raw VC that looks like this:
```

```json
vcDataModel: {
  "@context": [ "https://www.w3.org/2018/credentials/v1" ],
  type: [ "VerifiableCredential", "SanctionsCredential" ],
  id: "urn:uuid:daa5ff01-9b7d-485b-9410-22a42952d46c",
  issuer: "did:key:z6Mkn4w6nSaWe4fjNJRvaHZwFnMm5VexvjzDeozEu2G7jC34",
  issuanceDate: "2023-12-14T08:20:41Z",
  credentialSubject: {
    id: "did:key:z6MkjGSeekPGE9QfczHWyW8v2ZzJU68kqSHzV7L2dmQyuyDu",
    listsChecked: "FBI's Most Wanted, USA Watchlist, EU Watchlist"
  }
}
```

> To complete your basic VC issuance flow, you can now sign and return the credential. Signing will return a [VC JSON Web Token](/docs/web5/build/verifiable-credentials/jwt-to-vc#what-is-a-jwt), which is ideal for secure transmission of the credential:

Reviewer: ```suggestion
To complete the issuance, you must sign the credential. Signing will return a [VC JSON Web Token](/docs/web5/build/verifiable-credentials/jwt-to-vc#what-is-a-jwt), which is ideal for secure transmission of the credential:
```

```typescript
const signedVC = await vc.sign({did: issuerDid});
return res.status(201).json({ credential: signedVC });
```

> With that, you've implmented a basic API service to accept requests, process user data, and issue a VC!

Reviewer: ```suggestion
With that, you've implmented a basic credential issuance service to accept requests and issue VCs.
```

> ## Additional Resources

Reviewer: let's delete this section

Author: Are you referring to the "Additional Resources" Section?

Reviewer: yes

Author: Removed

# Onboarding

In a tbDEX network, Credential Issuers ensure trust between Wallets and PFIs by independently attesting claims about the party to whom they are issuing the [verifiable credential (VC)](/docs/web5/learn/verifiable-credentials).

Common examples of credentials that are necessary in financial transactions are ones that verify the identity of a Wallet owner.

Although credential issuance is not a concept that is exclusive to tbDEX, it's a key element in enabling compliant transactions.

In this guide, we'll cover the basic steps to set up your own Credential Issuance service.

## Background

Although you may know you want to build a service to issue Verifiable Credentails (VCs) to faciliate trust and transactions through tbDEX, you may be looking for more background knowledge before beginning development. Our [overview guide on VCs](/docs/web5/learn/verifiable-credentials) serves as background on the overall topic of issuance and VCs.

Additionally, developing an Issuer app will require use of our `@web5/credentials` package via `npm`, which we'll cover in more specifics later.

## Overview

> A credential issuer app runs as an API service that exposes endpoints where third parties can request credentials. You can see a full end-to-end implementation of a credential issuer's workflow [in our documentation here](/docs/web5/build/verifiable-credentials/fan-club-vc). For the sake of this section, we'll go over what you as a tbDEX issuer will need to set up your app:

Reviewer: this is still pointing to the Taylor Swift example. Please update this to point to the generic VC Issuance guide instead

Reviewer: I was pretty sure I wrote a suggestion for this paragraph but now I cant find it 😫 let me try again...

```suggestion
A Credential Issuer app runs as an API service that exposes endpoints where third parties can request credentials. 

If you already have an Issuer app up and running,  check out the guide on [how to issue credentials with Web5](/docs/web5/build/verifiable-credentials/vc-issuance/).

If you need to create a Credential Issuer app from scratch, continue on with this guide where we'll cover how to: 
```

> - Create a web server using a technology like ** [express](https://expressjs.com/) ** that will serve as an API endpoint.

Reviewer: ```suggestion
- Create a web server using a technology like ** [Express](https://expressjs.com/) ** that will serve as an API endpoint.
```

- Design the credential you'd like to offer.

> - Implement your issuance path, where you'll accept a request for issuance, perform checks required to verify the requester's information, and mint the credential.

Reviewer: ```suggestion
- Implement an issuance path where you'll accept a request for issuance, perform any necessary checks, and issue the credential.
```

- Return the credential to the requester.

## Environment Setup

To set up the VC issuance service, we'll use Express. Follow these instructions to [install Express](https://expressjs.com/en/starter/installing.html). 

You'll also need to install `@web5/credentials` to create and issue credentials:

```bash
npm i @web5/credentials
```

> Within your service's entry point file (e.g. index.js), import the following:

Reviewer: ```suggestion
Within your service's entry point file (e.g. main.ts), import the following:
```

> ```typescript title=index.ts

Reviewer: should this be `main`? that's why I asked if `main.ts` is your main entry point in that other comment. wanted to make sure I'm not introducing a new file (index) if it should be main

```suggestion
```typescript title=main.ts
```

In that code, you'll want to perform any checks required for you to issue them their credential which will later be verified by an interested 3rd party. In the example below, we'll assume that any user that makes the API request with a valid JWT token is who they say they are, and that the `verifyUserData()` method will then use the identity represented by the JWT token to check this individual against known sanctions lists.

```typescript title=main.ts
api.get('/credential', async (req, res) => {
    try {
        // Intake user data or send back any responses/redirects requesting more info
        const authHeader = req.headers['authorization'];

        if (!authHeader) {
            return res.status(401).json({ errors: ['Authorization header required'] });
        }

        const [_, compactJwt] = authHeader.split('Bearer ');
        const { signer, payload } = await JwtVerifier.verify(compactJwt);

        // Process & verify user data
        const listsChecked = verifyUserData(payload);

        if (listsChecked == null) {

>             res.status(401).json({

Reviewer: should this be a return statement? otherwise, it'll still create the credential after the fail, right?

```suggestion
            return res.status(401).json({
```
